packages/apollo-federation/src/composition/validate/sdl/possibleTypeExtensions.ts

```diff
@@ -0,0 +1,131 @@
+import { SDLValidationContext } from 'graphql/validation/ValidationContext';
+import {
+  ASTVisitor,
+  isObjectType,
+  isScalarType,
+  isInterfaceType,
+  isUnionType,
+  isEnumType,
+  isInputObjectType,
+  Kind,
+  // ObjectTypeExtensionNode,
+  isTypeDefinitionNode,
+  ObjectTypeExtensionNode,
+  InterfaceTypeExtensionNode,
+  // TypeExtensionNode,
+} from 'graphql';
+import { errorWithCode, logServiceAndType } from '../../utils';
+
+type FederatedExtensionNode = (
+  | ObjectTypeExtensionNode
+  | InterfaceTypeExtensionNode) & {
+  serviceName?: string | null;
+};
+
+export function PossibleTypeExtensions(
+  context: SDLValidationContext,
+): ASTVisitor {
+  const schema = context.getSchema();
+  const definedTypes = Object.create(null);
+
+  for (const def of context.getDocument().definitions) {
+    if (isTypeDefinitionNode(def)) {
+      definedTypes[def.name.value] = def;
+    }
+  }
+
+  const checkExtension = (node: FederatedExtensionNode) => {
+    const typeName = node.name.value;
+    const defNode = definedTypes[typeName];
+    const existingType = schema && schema.getType(typeName);
+
+    const serviceName = node.serviceName;
+    if (!serviceName) return;
+
+    if (defNode) {
+      const expectedKind = defKindToExtKind[defNode.kind];
+      const baseKind = defNode.kind;
+      if (expectedKind !== node.kind) {
+        context.reportError(
+          errorWithCode(
+            'EXTENSION_OF_WRONG_KIND',
+            logServiceAndType(serviceName, typeName) +
+              `\`${typeName}\` was originally defined as a ${baseKind} and can only be extended by a ${expectedKind}. ${serviceName} defines ${typeName} as a ${
+                node.kind
+              }`,
+          ),
+        );
+      }
+    } else if (existingType) {
+      const expectedKind = typeToExtKind(existingType);
+      const baseKind = typeToKind(existingType);
+      if (expectedKind !== node.kind) {
+        context.reportError(
+          errorWithCode(
+            'EXTENSION_OF_WRONG_KIND',
+            logServiceAndType(serviceName, typeName) +
+              `\`${typeName}\` was originally defined as a ${baseKind} and can only be extended by a ${expectedKind}. ${serviceName} defines ${typeName} as a ${
+                node.kind
+              }`,
+          ),
+        );
+      }
+    } else {
+      context.reportError(
+        errorWithCode(
+          'EXTENSION_WITH_NO_BASE',
+          logServiceAndType(serviceName, typeName) +
+            `\`${typeName}\` is an extension type, but \`${typeName}\` is not defined in any service`,
+        ),
+      );
+    }
+  };
+
+  return {
+    ObjectTypeExtension: checkExtension,
```

Comment:
do we want to support this for any other types?

Reply:
Not currently! In the future Unions may come into play, but we have a lot of thinking on that first 😆